core/domain/stats_services.py

```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+            this will be equal to 'question_id'.
+        entity_type: str. The type of entity i.e ENTITY_TYPE_EXPLORATION
+            or ENTITY_TYPE_QUESTION, which are declared in feconf.py.
+
+    Returns:
+        LearnerAnswerDetails. The learner answer domain object
+            or None if the model does not exist.
+    """
+    learner_answer_details_model = (
+        stats_models.LearnerAnswerDetailsModel.get_model_instance(
+            entity_type, state_reference))
+    if learner_answer_details_model:
+        learner_answer_details = get_learner_answer_details_from_model(
+            learner_answer_details_model)
+        return learner_answer_details
+    return None
+
+
+def save_learner_answer_details(
+        learner_answer_details, state_reference, entity_type):
+    """Saves the LearnerAnswerDetails domain object in the datatstore.
+
+    Args:
+        learner_answer_details: LearnerAnswerDetails. The  learner answer
+            details domain object which is to be saved.
+        state_reference: str. This is used to refer to a state
+            in an exploration or question. For an exploration the
+            value will be equal to 'exp_id.state_name' and for question
+            this will be equal to 'question_id'.
+        entity_type: str. The type of entity i.e ENTITY_TYPE_EXPLORATION
+            or ENTITY_TYPE_QUESTION, which are declared in feconf.py.
+    """
+    learner_answer_details.validate()
+    learner_answer_details_model = (
+        stats_models.LearnerAnswerDetailsModel.get_model_instance(
+            entity_type, state_reference))
+    instance_id = stats_models.LearnerAnswerDetailsModel.get_instance_id(
+        learner_answer_details.entity_type,
+        learner_answer_details.state_reference)
+    if learner_answer_details_model:
+        if learner_answer_details_model.id == instance_id:
+            learner_answer_details_model.state_reference = (
+                learner_answer_details.state_reference)
+            learner_answer_details_model.entity_type = (
+                learner_answer_details.entity_type)
+            learner_answer_details_model.interaction_id = (
+                learner_answer_details.interaction_id)
+            learner_answer_details_model.learner_answer_info_list = (
+                [learner_answer_info.to_dict() for learner_answer_info
+                 in learner_answer_details.learner_answer_info_list])
+            learner_answer_details_model.learner_answer_info_schema_version = (
+                learner_answer_details.learner_answer_info_schema_version)
+            learner_answer_details_model.accumulated_answer_info_json_size_bytes = ( #pylint: disable=line-too-long
+                learner_answer_details.accumulated_answer_info_json_size_bytes)
+            learner_answer_details_model.put()
+        else:
+            learner_answer_details_model.delete()
+            stats_models.LearnerAnswerDetailsModel.create_model_instance(
```

Comment:
Can you find a way to not duplicate this and lines 1130-1137?

Reply:
Done